### PR TITLE
Don't assume that base classes are already resolved.

### LINKED ISSUE #359: bindgen generates duplicate definitions
This command and input generates two distinct definitions for `std_char_traits`. I tried both `--enable-cxx_namespaces` and `--disable-name-namespacing`, but neither made a significant difference.

Command: `bindgen stdcxx.i -- -x c++ -std=c++14`

Input:
```
namespace std
{
  template < typename > struct char_traits;
    template < typename _CharT, typename _Traits =
    char_traits < _CharT > >class basic_fbstring;
}
namespace __gnu_cxx
{
  template < typename > struct char_traits;
}
namespace std
{
  template < class _CharT > struct char_traits:__gnu_cxx::char_traits <
    _CharT >
  {
  };
}
```

### PR DESCRIPTION
**Pull request by emilio**
Since it may not be the case.

Fixes #359 

Also fixes a few other test cases in the codebase that had the wrong namespace relationship.

r? @fitzgen

### CODE REVIEW

**Comment by emilio** [comment]
PSA stylo people (@heycam, @upsuper, @bholley, @Manishearth), this is likely to produce a more-noisy-than-usual bindings update, since we parse stuff in different order.

But I've verified this doesn't break anything in stylo.

File: libbindgen/src/lib.rs [hunk 1]
```diff
@@ -17,13 +18,13 @@
 #![allow(non_upper_case_globals)]
 
 #[macro_use]
+#[allow(unused_extern_crates)]
```
Comment (fitzgen) [inline libbindgen/src/lib.rs]: It still warns here despite `#[macro_use]` and our macro usage?

Comment (emilio) [inline libbindgen/src/lib.rs]: Yep, that's a known false positive from that lint.

**Pull request merged**